packages/lodestar/src/sync/range/chain.ts

```diff
@@ -0,0 +1,350 @@
+    const batches = toArr(this.batches);
+
+    // Only request batches up to the buffer size limit
+    // Note: Don't count batches in the AwaitingValidation state, to prevent stalling sync
+    // if the current processing window is contained in a long range of skip slots.
+    const batchesInBuffer = batches.filter((batch) => {
+      return batch.state.status === BatchStatus.Downloading || batch.state.status === BatchStatus.AwaitingProcessing;
+    });
+    if (batchesInBuffer.length > BATCH_BUFFER_SIZE) {
+      return null;
+    }
+
+    // This line decides the starting epoch of the next batch. MUST ensure no duplicate batch for the same startEpoch
+    const startEpoch = toBeDownloadedStartEpoch(batches, this.startEpoch, this.opts);
+
+    // Don't request batches beyond the target head slot
+    if (startEpoch > targetEpoch) {
+      return null;
+    }
+
+    const batch = new Batch(startEpoch, this.config, this.logger, this.opts);
+    if (this.batches.has(startEpoch)) throw Error(`Batch already ${startEpoch} exists`);
+    this.batches.set(startEpoch, batch);
+    return batch;
+  }
+
+  /**
+   * Requests the batch asigned to the given id from a given peer.
+   */
+  private async sendBatch(batch: Batch, peer: PeerId): Promise<void> {
+    // Inform the batch about the new request
+    batch.startDownloading(peer);
+
+    try {
+      const blocks = await this.downloadBeaconBlocksByRange(peer, batch.request);
+      batch.downloadingSuccess(blocks || []);
+
+      this.triggerBatchProcessor();
+    } catch (e) {
+      batch.downloadingError(e);
+    } finally {
+      // Pre-emptively request more blocks from peers whilst we process current blocks
+      this.triggerBatchDownloader();
+    }
+  }
+
+  /**
+   * Processes the next ready batch if any
+   */
+  private async processCompletedBatches(targetEpoch: Epoch): Promise<void> {
+    // ES6 Map guarantees insertion order so older batches are processed first
+    const batch = getNextBatchToProcess(toArr(this.batches));
+    if (batch) await this.processBatch(batch, targetEpoch);
+  }
+
+  /**
+   * Sends `batch` to the processor. Note: batch may be empty
+   */
+  private async processBatch(batch: Batch, targetEpoch: Epoch): Promise<void> {
+    try {
+      const blocks = batch.startProcessing();
+      await this.processChainSegment(blocks);
+      batch.processingSuccess();
+
+      // If the processed batch was not empty, we can validate previous unvalidated blocks.
+      if (blocks.length > 0) {
+        this.advanceChain(batch.startEpoch, targetEpoch);
+      }
+
+      // Potentially process next AwaitingProcessing batch
+      this.triggerBatchProcessor();
+    } catch (e) {
+      batch.processingError(e);
+
+      // At least one block was successfully verified and imported, so we can be sure all
+      // previous batches are valid and we only need to download the current failed batch.
+      if (e instanceof BlockProcessorError && e.importedBlocks.length > 0) {
```

Comment:
I don't see this BlockProcessorError being thrown anywhere

Reply:
Right, I'm waiting for Cayman to tweak the chain segment processor and add this error. The error is not necessary to work, it just adds a performance boost in some situations

Reply:
looks like we'll use `ChainSegmentError`? https://github.com/ChainSafe/lodestar/blob/master/packages/lodestar/src/chain/errors/blockError.ts#L125

Reply:
will change in a separate PR